### ISSUE TITLE
training llama2 by replacing attn with xformer got loss exposion

### ISSUE THREAD

**Issue by lucasjinreal** [facebookresearch/xformers#815]
Hi, to save mem when training, I replaced attn with xformers:

```python
def forward(
    self,
    hidden_states: torch.Tensor,
    attention_mask: Optional[torch.Tensor] = None,
    position_ids: Optional[torch.Tensor] = None,
    past_key_value: Optional[Tuple[torch.Tensor]] = None,
    output_attentions: bool = False,
    use_cache: bool = False,
) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
    bsz, q_len, _ = hidden_states.size()

    query_states = (
        self.q_proj(hidden_states)
        .view(bsz, q_len, self.num_heads, self.head_dim)
        .transpose(1, 2)
    )
    key_states = (
        self.k_proj(hidden_states)
        .view(bsz, q_len, self.num_heads, self.head_dim)
        .transpose(1, 2)
    )
    value_states = (
        self.v_proj(hidden_states)
        .view(bsz, q_len, self.num_heads, self.head_dim)
        .transpose(1, 2)
    )

    kv_seq_len = key_states.shape[-2]
    if past_key_value is not None:
        kv_seq_len += past_key_value[0].shape[-2]
    cos, sin = self.rotary_emb(value_states, seq_len=kv_seq_len)
    query_states, key_states = apply_rotary_pos_emb(
        query_states, key_states, cos, sin, position_ids
    )
    # [bsz, nh, t, hd]

    if past_key_value is not None:
        # reuse k, v, self_attention
        key_states = torch.cat([past_key_value[0], key_states], dim=2)
        value_states = torch.cat([past_key_value[1], value_states], dim=2)

    past_key_value = (key_states, value_states) if use_cache else None

    # Copied from modeling_open_llama.py
    attn_weights = None

    query_states = query_states.transpose(1, 2)
    key_states = key_states.transpose(1, 2)
    value_states = value_states.transpose(1, 2)
    attn_output = xops.memory_efficient_attention(
        query_states,
        key_states,
        value_states,
        attn_bias=xops.LowerTriangularMask(),
        p=0,
    )
    attn_output = attn_output.reshape(bsz, q_len, self.hidden_size)

    attn_output = self.o_proj(attn_output)

    assert not use_cache, "xformer does not support use_cache=True"

    if not output_attentions:
        attn_weights = None

    return attn_output, attn_weights, past_key_value


def replace_llama_attn_with_xformer():
    transformers.models.llama.modeling_llama.LlamaAttention.forward = forward
```

When training on llama2 7B, but everytime it training the loss are explosion, disable xformers, then it works normal.

anybody got some idea?

**Comment by danthe3rd** [comment]
Hi,
You should get roughly the same loss with xformers/without xformers (in fact the team at Meta who trained LLaMa v2 used xformers). This means there is something wrong in how you use xformers.
Now I'm not exactly sure what is wrong in the code, but if I were you I would start by testing this layer individually, and check if the output of a single layer matches the reference layer.
